src/libraries/Microsoft.Bcl.TimeProvider/src/System/Threading/Tasks/TimeProviderTaskExtensions.cs

```diff
@@ -218,5 +218,46 @@ public static async Task<TResult> WaitAsync<TResult>(this Task<TResult> task, Ti
             return task.Result;
         }
 #endif // NET8_0_OR_GREATER
+
+        /// <summary>Initializes a new instance of the <see cref="CancellationTokenSource"/> class that will be canceled after the specified <see cref="TimeSpan"/>. </summary>
+        /// <param name="timeProvider">The <see cref="TimeProvider"/> with which to interpret the <paramref name="delay"/>. </param>
+        /// <param name="delay">The time interval to wait before canceling this <see cref="CancellationTokenSource"/>. </param>
+        /// <exception cref="ArgumentOutOfRangeException"> The <paramref name="delay"/> is negative and not equal to <see cref="Timeout.InfiniteTimeSpan" /> or greater than maximum allowed timer duration.</exception>
+        /// <returns><see cref="CancellationTokenSource"/> that will be canceled after the specified <paramref name="delay"/>.</returns>
+        /// <remarks>
+        /// The countdown for the delay starts during the call to the constructor. When the delay expires,
+        /// the constructed <see cref="CancellationTokenSource"/> is canceled if it has
+        /// not been canceled already.
+        /// If running on framework version prior to .NET 8.0, there is a constraint when invoking <see cref="CancellationTokenSource.CancelAfter(TimeSpan)"/> on the resultant object.
+        /// This action will not terminate the initial timer indicated by <paramref name="delay"/>. However, this restriction does not apply on .NET 8.0 and later versions.
+        /// </remarks>
+        public static CancellationTokenSource CreateCancellationTokenSource(this TimeProvider timeProvider, TimeSpan delay)
+        {
+#if NET8_0_OR_GREATER
+            return new CancellationTokenSource(delay, timeProvider);
+#else
+            if (timeProvider is null)
+            {
+                throw new ArgumentNullException(nameof(timeProvider));
+            }
+
+            if (delay != Timeout.InfiniteTimeSpan && delay < TimeSpan.Zero)
+            {
+                throw new ArgumentOutOfRangeException(nameof(delay));
+            }
+
+            var cts = new CancellationTokenSource();
+
+            if (timeProvider == TimeProvider.System)
+            {
+                cts.CancelAfter(delay);
+                return cts;
+            }
+
+            ITimer timer = timeProvider.CreateTimer(s => ((CancellationTokenSource)s).Cancel(), cts, delay, Timeout.InfiniteTimeSpan);
```

Comment:
```suggestion
            ITimer timer = timeProvider.CreateTimer(s =>
            {
                try
                {
                    ((CancellationTokenSource)s).Cancel();
                }
                catch (ObjectDisposedException) { }
            }, cts, delay, Timeout.InfiniteTimeSpan);
```
Please also add a test that:
1. Creates the CTS with a delay that we control via the time provider.
2. Disposes the CTS.
3. Then fires the timer.  We want to validate the process doesn't crash.